core/src/main/java/org/elasticsearch/common/util/ExtensionPoint.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.util;
+
+import org.elasticsearch.common.inject.Binder;
+import org.elasticsearch.common.inject.multibindings.MapBinder;
+import org.elasticsearch.common.inject.multibindings.Multibinder;
+import org.elasticsearch.common.settings.Settings;
+
+import java.util.*;
+
+/**
+ * This class defines an official elasticsearch extension point. It registers
+ * all extensions by a single name and ensures that extensions are not registered
+ * more than once.
+ */
+public abstract class ExtensionPoint<T> {
+    protected final String name;
+    protected final Class<T> extensionClass;
+    protected final Set<String> buildIns;
+    protected final Class<?> accessor;
+
+    public ExtensionPoint(String name, Class<T> extensionClass, Set<String> buildIns, Class<?> accessor) {
+        this.name = name;
+        this.extensionClass = extensionClass;
+        this.buildIns = buildIns;
+        this.accessor = accessor;
+    }
+
+    public final void bind(Binder binder) {
+        if (accessor == null) {
+            throw new IllegalStateException("Can't bind null accessor");
+        }
+        binder.bind(accessor).asEagerSingleton();
+        bindExtensions(binder);
+        onBind(binder);
+    }
+
+    protected abstract void bindExtensions(Binder binder);
+
+    protected void onBind(Binder binder) {
+    }
+
+
+    public static class MapExtensionPoint<T> extends ExtensionPoint<T> {
```

Comment:
Personally I would do these as two separate classes. Then the ctor here can take a map of builtins instead of set of names (which relies on registering those names independently)?